From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 05-05

  Description: kernel: Incomplete stack traces.
  Symptom:     Incomplete stack traces in case of bugs.
  Problem:     show_stack() passes a to the current stack frame to show_trace().
               Because of tail call optimization the pointer doesn't point to
               the original stack frame anymory and therefore traces are
               wrong.
  Solution:    Don't pass the pointer of the current stack frame to show_trace().
               Instead let show_trace() calculate the pointer on its own.
  Problem-ID:  -
  Archs:       s390-31, s390-64

Acked-by: Ihno Krumreich <ihno@suse.de>

--- linux-2.5/arch/s390/kernel/traps.c	20 Mar 2006 15:56:34 -0000	1.51
+++ linux-2.5/arch/s390/kernel/traps.c	2 Jun 2006 14:10:56 -0000	1.51.2.1
@@ -150,13 +150,11 @@
 	unsigned long *stack;
 	int i;
 
-	// debugging aid: "show_stack(NULL);" prints the
-	// back trace for this cpu.
-
 	if (!sp)
-		sp = task ? (unsigned long *) task->thread.ksp : __r15;
+		stack = task ? (unsigned long *) task->thread.ksp : __r15;
+	else
+		stack = sp;
 
-	stack = sp;
 	for (i = 0; i < kstack_depth_to_print; i++) {
 		if (((addr_t) stack & (THREAD_SIZE-1)) == 0)
 			break;

